File: docs/perry-main-src/perry-main/benchmarks/json_polyglot/bench.go
// JSON parse + stringify polyglot benchmark — Go.
// 10k records, ~1 MB blob, 50 iterations.
// IDENTICAL workload to bench.ts / bench.rs / bench.swift / bench.cpp / bench.js.

package main

import (
	"encoding/json"
	"fmt"
	"strconv"
	"time"
)

type Nested struct {
	X int `json:"x"`
	Y int `json:"y"`
}

type Item struct {
	Id     int      `json:"id"`
	Name   string   `json:"name"`
	Value  float64  `json:"value"`
	Tags   []string `json:"tags"`
	Nested Nested   `json:"nested"`
}

func main() {
	items := make([]Item, 10000)
	for i := 0; i < 10000; i++ {
		items[i] = Item{
			Id:    i,
			Name:  "item_" + strconv.Itoa(i),
			Value: float64(i) * 3.14159,
			Tags:  []string{"tag_" + strconv.Itoa(i%10), "tag_" + strconv.Itoa(i%5)},
			Nested: Nested{X: i, Y: i * 2},
		}
	}
	blob, _ := json.Marshal(items)

	// Warmup
	for i := 0; i < 3; i++ {
		var parsed []Item
		_ = json.Unmarshal(blob, &parsed)
		_, _ = json.Marshal(parsed)
	}

	const iterations = 50
	start := time.Now()

	checksum := 0
	for iter := 0; iter < iterations; iter++ {
		var parsed []Item
		_ = json.Unmarshal(blob, &parsed)
		checksum += len(parsed)
		reStringified, _ := json.Marshal(parsed)
		checksum += len(reStringified)
	}

	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("ms:%d\n", elapsed)
	fmt.Printf("checksum:%d\n", checksum)
}

File: docs/perry-main-src/perry-main/benchmarks/json_polyglot/bench.rs
// JSON parse + stringify polyglot benchmark — Rust (serde_json).
// 10k records, ~1 MB blob, 50 iterations.
// IDENTICAL workload to bench.ts / bench.go / bench.swift / bench.cpp / bench.js.
//
// To build: a Cargo.toml is provided alongside this file; run.sh invokes
// `cargo build --release` (or `--profile release-aggressive` with the
// flags table from benchmarks/README.md).

use serde::{Deserialize, Serialize};
use std::time::Instant;

#[derive(Serialize, Deserialize)]
struct Nested {
    x: i64,
    y: i64,
}

#[derive(Serialize, Deserialize)]
struct Item {
    id: i64,
    name: String,
    value: f64,
    tags: Vec<String>,
    nested: Nested,
}

fn main() {
    let mut items: Vec<Item> = Vec::with_capacity(10_000);
    for i in 0..10_000i64 {
        items.push(Item {
            id: i,
            name: format!("item_{}", i),
            value: (i as f64) * 3.14159,
            tags: vec![format!("tag_{}", i % 10), format!("tag_{}", i % 5)],
            nested: Nested { x: i, y: i * 2 },
        });
    }
    let blob = serde_json::to_string(&items).expect("stringify blob");

    // Warmup
    for _ in 0..3 {
        let parsed: Vec<Item> = serde_json::from_str(&blob).expect("parse");
        let _ = serde_json::to_string(&parsed).expect("stringify");
    }

    const ITERATIONS: usize = 50;
    let start = Instant::now();

    let mut checksum: usize = 0;
    for _ in 0..ITERATIONS {
        let parsed: Vec<Item> = serde_json::from_str(&blob).expect("parse");
        checksum += parsed.len();
        let re_stringified = serde_json::to_string(&parsed).expect("stringify");
        checksum += re_stringified.len();
    }

    let elapsed = start.elapsed().as_millis();
    println!("ms:{}", elapsed);
    println!("checksum:{}", checksum);
}
